[SymmMem] Add NVSHMEM sync_all support into Triton

codingwithsurya opened the pull request:
Adds `sync_all()` function for local store visibility synchronization in NVSHMEM Triton kernels. Provides memory ordering for local operations without remote completion guarantees.

Tests: `python test/distributed/test_nvshmem_triton.py -k test_triton_sync`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #158718
* #158676
* #158515
* #158514
* #158513
* __->__ #158512
* #158511



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

codingwithsurya committed:
[SymmMem] ADD NVSHMEM sync_all support into Triton

[ghstack-poisoned]

codingwithsurya committed:
Update on "[SymmMem] Add NVSHMEM sync_all support into Triton"

Adds `sync_all()` function for local store visibility synchronization in NVSHMEM Triton kernels. Provides memory ordering for local operations without remote completion guarantees.

Tests: `python test/distributed/test_nvshmem_triton.py -k test_triton_sync`




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

codingwithsurya committed:
Update on "[SymmMem] Add NVSHMEM sync_all support into Triton"

Adds `sync_all()` function for local store visibility synchronization in NVSHMEM Triton kernels. Provides memory ordering for local operations without remote completion guarantees.

Tests: `python test/distributed/test_nvshmem_triton.py -k test_triton_sync`




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

codingwithsurya committed:
Update on "[SymmMem] Add NVSHMEM sync_all support into Triton"

Adds `sync_all()` function for local store visibility synchronization in NVSHMEM Triton kernels. Provides memory ordering for local operations without remote completion guarantees.

Tests: `python test/distributed/test_nvshmem_triton.py -k test_triton_sync`




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

codingwithsurya committed:
Update on "[SymmMem] Add NVSHMEM sync_all support into Triton"

Adds `sync_all()` function for local store visibility synchronization in NVSHMEM Triton kernels. Provides memory ordering for local operations without remote completion guarantees.

Tests: `python test/distributed/test_nvshmem_triton.py -k test_triton_sync`




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented on torch/distributed/_symmetric_memory/_nvshmem_triton.py:
```diff
@@ -230,3 +230,14 @@ def barrier_all(_builder=None):  # type: ignore[no-untyped-def]
+            "",
+            "",
+            [],
+            {(): ("nvshmem_sync_all", core.dtype("int32"))},
```
maybe a naive question, it looks very similar to `nvshmem_barrier_all`, what is the difference here? And when should user use one or the other can you maybe add some comments to elaborate that? Thanks

codingwithsurya replied:
`nvshmem_barrier_all()` blocks until all PEs call it and ensures completion of both local stores **and** remote memory updates (e.g., puts, AMOs). `nvshmem_sync_all()` only ensures completion of local stores, not remote updates. 

you want to use barrier_all when you need remote memory visibility guarantees and sync_all as a lighter-weight barrier if you’ll handle remote completion separately (like via `nvshmem_quiet()`).

I added a comment [here](https://github.com/pytorch/pytorch/pull/158512/files#diff-2746214da939401c56c7499252c05299ed7f78198c3f340774df4cfa1ac89a6dR203) that showed difference but maybe i should make it a little more detailed. Will update the commit.

fduwjj reviewed (approved):
In general this looks good to me.

codingwithsurya committed:
Update on "[SymmMem] Add NVSHMEM sync_all support into Triton"

Adds `sync_all()` function for local store visibility synchronization in NVSHMEM Triton kernels. Provides memory ordering for local operations without remote completion guarantees.

Tests: `python test/distributed/test_nvshmem_triton.py -k test_triton_sync`




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]
